# 1. Simple Route to Greet the User

In [ ]:
# !pip install flask

In [1]:
from flask import Flask

app = Flask(__name__)

@app.route('/')
def greet():
    return 'Hello, World!'

if __name__ == '__main__':
    app.run(debug=False)


### Accessing the endpoint via curl (Commandline / Terminal)
# curl http://localhost:5000/

 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [30/Oct/2024 17:21:10] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [30/Oct/2024 17:21:11] "GET /favicon.ico HTTP/1.1" 404 -
127.0.0.1 - - [30/Oct/2024 17:21:41] "GET / HTTP/1.1" 200 -


# 2. Route with User Input and Computation
- Let's create a route that takes a number as input, squares it, and displays the result.
- Also `/greet/<name>` route to your Flask application, which takes the username from the URL and greets them.

- We use `render_template_string` to render an HTML template with the result.
- `{{variable}}` syntax is from `Jinja2`, used for templating.

In [3]:
from flask import Flask, request, render_template_string

app = Flask(__name__)

# Captures a string name from the URL.
@app.route('/greet/<name>')
def greet_user(name):
    return render_template_string('<h1>Hello, {{name}}!</h1>', name=name)

# The route /square/<int:number> captures an integer from the URL.
@app.route('/square/<int:number>')
def square(number):
    result = number ** 2
    return render_template_string('<h1>{{number}} squared is {{result}}</h1>', number=number, result=result)

if __name__ == '__main__':
    app.run(debug=False)

    
# Testing with curl
# Replace <name> with any username you like.
# curl http://localhost:5000/greet/Prashant
               
# curl http://localhost:5000/square/5    



 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [30/Oct/2024 17:23:02] "GET / HTTP/1.1" 404 -
127.0.0.1 - - [30/Oct/2024 17:23:39] "GET /greet/Prashant HTTP/1.1" 200 -
127.0.0.1 - - [30/Oct/2024 17:23:39] "GET /favicon.ico HTTP/1.1" 404 -
127.0.0.1 - - [30/Oct/2024 17:23:46] "GET /greet/Amy HTTP/1.1" 200 -
127.0.0.1 - - [30/Oct/2024 17:24:18] "GET /square/100 HTTP/1.1" 200 -
127.0.0.1 - - [30/Oct/2024 17:24:38] "GET /square/5 HTTP/1.1" 200 -
127.0.0.1 - - [30/Oct/2024 17:24:42] "GET /square/ HTTP/1.1" 404 -
127.0.0.1 - - [30/Oct/2024 17:25:31] "GET /greet/Prashant%20Sahu HTTP/1.1" 200 -
127.0.0.1 - - [30/Oct/2024 17:26:59] "GET /greet/John%20Doe HTTP/1.1" 200 -


**Additional Notes:**

- **URL Encoding:**
  - If the name contains spaces or special characters, ensure it's URL-encoded.
  - Refer to: https://www.w3schools.com/tags/ref_urlencode.asp
  - Online URL Encoder: https://www.urlencoder.io/
  - For example, for "John Doe":

    ```bash
    curl http://localhost:5000/greet/John%20Doe
    ```

- **Error Handling:**
  - Currently, the code does not handle errors like missing parameters or invalid inputs.
  - You might want to add error handling to make your application more robust.

**Feel free to ask if you need further clarification or assistance!**

# 3. Route Demonstrating the GET Method
Let's create a route that accepts query parameters using the GET method.

- `400` is the HTTP status code for `Bad Request`.
- By returning `400`, the server indicates to the client that the request was invalid due to missing required parameters.

In [7]:
from flask import Flask, request, render_template_string
import math

app = Flask(__name__)

@app.route('/sqrt', methods=['GET'])
def compute_sqrt():
    try:
        # Retrieve numbers from query parameters and convert to float
        num1 = request.args.get('num1', type=float)
        num2 = request.args.get('num2', type=float)

        # Check if both numbers are provided
        if num1 is None or num2 is None:
            # Return error message with HTTP 400 status code
            return render_template_string(
                '<h1>Error: Both num1 and num2 parameters are required.</h1>'
            ), 400

        # Prevent division by zero
        if num2 == 0:
            return render_template_string(
                '<h1>Error: Division by zero is not allowed.</h1>'
            ), 400

        # Compute the division of num1 by num2
        division_result = num1 / num2

        # Ensure the result is non-negative for square root
        if division_result < 0:
            return render_template_string(
                '<h1>Error: Cannot compute square root of a negative number.</h1>'
            ), 400

        # Compute the square root
        sqrt_result = math.sqrt(division_result)

        # Return the result
        return render_template_string(
            '<h1>Square root of {{division_result}} is {{sqrt_result}}</h1>',
            division_result=division_result,
            sqrt_result=sqrt_result
        )

    except ValueError:
        # Handle cases where inputs are not numbers
        return render_template_string(
            '<h1>Error: Invalid input. Please provide numeric values for num1 and num2.</h1>'
        ), 400

   
if __name__ == '__main__':
    app.run(debug=False)
    
# Testing with curl
# curl "http://localhost:5000/sqrt?num1=16&num2=4"
# curl "http://localhost:5000/sqrt?num2=4"
# curl "http://localhost:5000/sqrt?num1=16&num2=0"
# curl "http://localhost:5000/sqrt?num1=-16&num2=4"
# curl "http://localhost:5000/sqrt?num1=abc&num2=4"



 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [30/Oct/2024 17:30:22] "GET /sqrt?num1=16&num2=4 HTTP/1.1" 200 -
127.0.0.1 - - [30/Oct/2024 17:30:33] "GET /sqrt?num1=16 HTTP/1.1" 400 -
127.0.0.1 - - [30/Oct/2024 17:30:51] "GET /sqrt?num1=16&num2=0 HTTP/1.1" 400 -
127.0.0.1 - - [30/Oct/2024 17:31:00] "GET /sqrt?num1=-16&num2=4 HTTP/1.1" 400 -
127.0.0.1 - - [30/Oct/2024 17:31:36] "GET /sqrt?num1=16&num2=4 HTTP/1.1" 200 -


### Accessing via Web Browser:
You can also test the endpoint by navigating to:

- Valid Input: http://localhost:5000/sqrt?num1=16&num2=4
- Missing Parameters: http://localhost:5000/sqrt?num1=16
- Division by Zero: http://localhost:5000/sqrt?num1=16&num2=0
- Negative Result: http://localhost:5000/sqrt?num1=-16&num2=4

# 4. Route Demonstrating the POST Method
We'll create a route that accepts data via POST and displays it.

In [13]:
from flask import Flask, request, render_template_string

app = Flask(__name__)

@app.route('/submit', methods=['GET', 'POST'])
def submit():
    if request.method == 'POST':
        message = request.form.get('message', 'No message')
        return render_template_string('<h1>You submitted:<hr /> {{message}}</h1>', message=message)
    else:
        # Display a simple form
        return '''
            <form method="post" action="/submit">
                <input type="text" name="message" placeholder="Enter a message" />
                <input type="submit" />
            </form>
        '''

if __name__ == '__main__':
    app.run(debug=False)

    
# Access it here: http://127.0.0.1:5000/submit

# curl -X POST -d "your message" http://localhost:5000/submit

 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [30/Oct/2024 17:35:21] "POST /submit HTTP/1.1" 200 -
127.0.0.1 - - [30/Oct/2024 17:35:24] "GET /submit HTTP/1.1" 200 -
127.0.0.1 - - [30/Oct/2024 17:35:52] "POST /submit HTTP/1.1" 200 -


In [ ]:
# import requests

# # URL of the Flask endpoint
# url = 'http://127.0.0.1:5000/submit'

# # Data to be sent in the POST request
# data = {
#     'message': 'Hello, Flask!'
# }

# # Make the POST request
# response = requests.post(url, data=data)

# # Print the response content
# print(response.text)

### 4A: Modify your Flask endpoint to accept JSON data

In [15]:
from flask import Flask, request, jsonify

app = Flask(__name__)

@app.route('/submit', methods=['GET', 'POST'])
def submit():
    if request.method == 'POST':
        data = request.get_json()
        message = data.get('message', 'No message')
        return jsonify({'response': f'You submitted: {message}'})
    else:
        return '''
            <form method="post" action="/submit">
                <input type="text" name="message" placeholder="Enter a message" />
                <input type="submit" />
            </form>
        '''

if __name__ == '__main__':
    app.run(debug=False)


 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [30/Oct/2024 17:38:50] "POST /submit HTTP/1.1" 200 -


In [ ]:
# import requests

# url = 'http://127.0.0.1:5000/submit'

# json_data = {
#     'message': 'Hello, Flask with JSON!'
# }

# response = requests.post(url, json=json_data)

# print(response.json())


# 5. Advanced APP Route using HTML templates and CSS rendering

In [ ]:
from flask import Flask, request, render_template
import math

app = Flask(__name__)

@app.route('/')
def home():
    return render_template('home.html')

@app.route('/sqrt', methods=['GET','POST'])
def compute_sqrt():
    if request.method == 'POST':
        try:
            num1 = request.form.get('num1', type=float)
            num2 = request.form.get('num2', type=float)

            if num1 is None or num2 is None:
                error = 'Both Number 1 and Number 2 are required.'
                return render_template('form.html', error=error)

            if num2 == 0:
                error = 'Division by zero is not allowed.'
                return render_template('form.html', error=error)

            division_result = num1 / num2

            if division_result < 0:
                error = 'Cannot compute square root of a negative number.'
                return render_template('form.html', error=error)

            sqrt_result = math.sqrt(division_result)

            return render_template(
                'result.html',
                division_result=division_result,
                sqrt_result=sqrt_result
            )

        except (ValueError, TypeError):
            error = 'Invalid input. Please provide numeric values for Number 1 and Number 2.'
            return render_template('form.html', error=error)
    else:
        return render_template('form.html')

if __name__ == '__main__':
    app.run(debug=False)

# http://127.0.0.1:5000


 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [30/Oct/2024 17:41:57] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [30/Oct/2024 17:41:57] "GET /static/styles.css HTTP/1.1" 200 -
127.0.0.1 - - [30/Oct/2024 17:42:04] "GET /sqrt HTTP/1.1" 200 -
127.0.0.1 - - [30/Oct/2024 17:42:04] "GET /static/styles.css HTTP/1.1" 304 -
127.0.0.1 - - [30/Oct/2024 17:42:20] "POST /sqrt HTTP/1.1" 200 -
127.0.0.1 - - [30/Oct/2024 17:42:20] "GET /static/styles.css HTTP/1.1" 304 -
127.0.0.1 - - [30/Oct/2024 17:42:27] "GET /sqrt HTTP/1.1" 200 -
127.0.0.1 - - [30/Oct/2024 17:42:27] "GET /static/styles.css HTTP/1.1" 304 -
127.0.0.1 - - [30/Oct/2024 17:42:34] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [30/Oct/2024 17:42:34] "GET /static/styles.css HTTP/1.1" 304 -
127.0.0.1 - - [30/Oct/2024 17:42:38] "GET /sqrt HTTP/1.1" 200 -
127.0.0.1 - - [30/Oct/2024 17:42:38] "GET /static/styles.css HTTP/1.1" 304 -
127.0.0.1 - - [30/Oct/2024 17:42:50] "POST /sqrt HTTP/1.1" 200 -
127.0.0.1 - - [30/Oct/2024 17:42:50] "G

- app.py: The main Flask application script.
- static/: Contains static files like CSS, JavaScript, images, etc. The styles.css file is placed here.

- templates/: Contains HTML templates rendered by Flask.

- form.html is the template for the form page.
- result.html is the template for displaying the computation result.

## **Understanding the `304 Not Modified` Status Code**

### **What is `304 Not Modified`?**

- **Definition:**
  - The `304 Not Modified` is an HTTP status code indicating that the requested resource has not been modified since the last time it was requested.
- **Purpose:**
  - It tells the client (usually a web browser) that it can use the cached version of the resource, reducing the need to download it again.


### **Benefits of `304 Not Modified`:**

- **Reduced Server Load:**
  - Less data is sent over the network.
- **Faster Load Times:**
  - The browser can render the page more quickly using cached resources.
- **Optimized Bandwidth Usage:**
  - Particularly beneficial for users with limited or metered internet connections.

